### DISCUSSION TITLE
Static Maps integration for ZEIR

### DISCUSSION THREAD

**Discussion by f-odhiambo** [opensrp/fhircore#2687] [Ideas]
**User Story**
Static Maps will be available as PDF documents. 

**Accessing/Downloading the PDF**
These docs can either be 
1. Added to the HAPI FHIR server and synced to the app embedded as base64 encoded Binary resource and then decoded in the app (Larger LOE) - launch pdf `offline` 
2. or downloaded directly from a web server URL  -  requires the user to be `online` 

**Viewing the PDF**
From a product perspective, we can use approach 2 above 
1. Add an Overflow menu to open the static map 
2. Add a config to pass the URL and then add an action config to launch an intent to the browser to open the link to the static map pdf. _This will involve taking the user away from the app._ WIll need to check in with the design team whether this UX is acceptable 
3. Alternatively, we can figure out a way to add a view that can render the PDF from the provided link and then close it once you either back press or have an overlay button to do so  - _Needs RnD_

**Comment by pld** [comment]
So far the plan was to handle this with this roadmap item, which does not involve a Questionnaire, https://www.notion.so/onaio/8a1d2a62bfe94535b38e476aa2d54708?v=3c5467c43d44413dacd1f1e6062703b2&p=b2fc3ebe70284ec2b58285a9cffb0e8c&pm=s

**Reply by f-odhiambo** [reply]
Noted I had read something close to it on the thread

**Comment by f-odhiambo** [comment]
@rowo @HenryRae Kindly share your input on the UX for how we redirect users to access the static maps and back to the app when they need to 
CC @samkanga

**Reply by pld** [reply]
If we are going with option (2) I don't think there are any UX considerations here

**Comment by rowo** [comment]
Apple iphone OS has a special "app back button" whenever an app throws you to another app you can tap. Does Android provide the same? 

I don't remember the original workflow from ZEIR for this. Do you have that handy @f-odhiambo ? I can try to find it if not. 

In terms of it being satisfactory, this might be more a client discussion. We need something similar in Engage (open up a link to a website that's opened in the phone's default web browser).

**Reply by f-odhiambo** [reply]
@rowo Please do share more on this

**Comment by f-odhiambo** [comment]
In terms of general implementation  (Online case)
1. Add an AppConfig to `LAUNCH_URL` which can be used to  launch an intent to open the URL 
2. Update the overflow menu config to be able to accommodate this config 
3. The link can also be passed hardcoded or from a  rule from a sync resource 

Something like 
````json
    {
      "title": "Show user Manual",
      "titleColor": "#FF0000",
      "visible": "true",
      "enabled":  "true",
      "showSeparator": "true",
      "actions": [
        {
          "trigger": "ON_CLICK",
          "workflow": "LAUNCH_URL",
          "url": {
            "link": "https://www.google.com/user_manual.pdf",
            "redirect": "true"
          }
        }
      ]
    },
````

or 


Rule 
````json
    {
      "name": "content_link",
      "condition": "true",
      "actions": [
        "data.put('content_link', fhirPath.extractValue(SomeReource, \"SomeReource.name[0].select(url)\"))"
      ]
    },
````

Menu Item
````json
    {
      "title": "Show user Manual",
      "titleColor": "#FF0000",
      "visible": "true",
      "enabled":  "true",
      "showSeparator": "true",
      "actions": [
        {
          "trigger": "ON_CLICK",
          "workflow": "LAUNCH_URL",
          "url": {
            "link": "@{content_link}",
            "redirect": "true"
          }
        }
      ]
    },
````

**Reply by f-odhiambo** [reply]
**Offline user case** 
If we are to take an offline approach will need to figure out how to best store the PDF and launch it on the app

**Reply by pld** [reply]
I thought it was ok not to do that, we'd give instructions that you have to download the PDF via the link before going offline

**Comment by dubdabasoduba** [comment]
@f-odhiambo the above solution makes sense. I have a question and probably a new suggestion.  

**Question**
- Does this mean the user always has to the online? 

**New suggestion**
- Have the static PDFs added to the [DocumentReference](https://www.hl7.org/fhir/R4B/documentreference.html) resource
- Add the DocumentReference reference to the Composition resource so it's downloaded to the app.  
    - We can also store this using the [Knowledge Management](https://github.com/google/android-fhir/tree/master/knowledge) module on the Android SDK
    - This means the PDFs can be updated and re-downloaded by logging out and in. 
- Create a provision to open this resource from the app.  
    - Add a PDF viewer on the app. I think we had something on the WCARO apps.

**Reply by f-odhiambo** [reply]
Thanks, @dubdabasoduba I think this addresses the Offline approach, and also note that if we make changes then  we need a fresh install to resync the [DocumentReference](https://www.hl7.org/fhir/R4B/documentreference.html) resource if it's part of the config

**Reply by dubdabasoduba** [reply]
@f-odhiambo we do not need a fresh install, we just need a log out and login

**Reply by f-odhiambo** [reply]
Super :)

**Reply by f-odhiambo** [reply]
Now we need a PDF viewer to complete that route

**Reply by dubdabasoduba** [reply]
Agreed! that we need to build but we can copy from one of the OpenSRP v1 apps.

**Comment by f-odhiambo** [comment]
The need as of Sept 1st is to get clear requirements on both WDF and Colombia on how the user needs to interact with this requirement
@rowo @Gental-Giant 

CC @pld @dubdabasoduba @ageryck @allan-on

**Comment by rowo** [comment]
Btw, this is the issue for Columbia Engage: https://github.com/onaio/fhir-resources/issues/2279. It is incomplete though.

**Comment by samkanga** [comment]
@f-odhiambo, @pld  adding comments from PATH on this feature:
1. The client desires to have the maps open from within the App.
2. The client also desires to be able to open the downloaded static maps from the downloads folder in the event they want to use it away from the App.
3. The main use case for the static maps is micro-planning.

**Reply by dubdabasoduba** [reply]
This is interesting @pld we can do some R&D on whether using the knowledge manager can help us achieve both points 1 and 2.

**Reply by dubdabasoduba** [reply]
Point 1 Will mean we build a view on the app UI to read PDFs

**Comment by rowo** [comment]
On point 1, for the Columbia scenario, it makes sense to do it outside of the app so the client can switch and refer to it while in a counseling session. Otherwise, if inside the app, it would have to be launchable via a form (without exiting the form).
